daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala

```diff
@@ -24,21 +22,84 @@ sealed abstract class MatchingStringModule {
   //  * https://github.com/digital-asset/daml/pull/983#discussion_r282513324
   //  * https://github.com/scala/bug/issues/9565
   val Array: ArrayFactory[T]
+
+  def toStringMap[V](map: Map[T, V]): Map[String, V]
 }
 
-object MatchingStringModule extends (Regex => MatchingStringModule) {
+object MatchingStringModule {
 
-  override def apply(regex: Regex): MatchingStringModule = new MatchingStringModule {
+  def apply(string_regex: String): StringModule = new StringModule {
     type T = String
 
+    private val regex = string_regex.r
     private val pattern = regex.pattern
 
-    def fromString(s: String): Either[String, T] =
+    override def fromString(s: String): Either[String, T] =
       Either.cond(pattern.matcher(s).matches(), s, s"""string "$s" does not match regex "$regex"""")
 
-    def equalInstance: Equal[T] = scalaz.std.string.stringInstance
+    override def equalInstance: Equal[T] = scalaz.std.string.stringInstance
+
+    override val Array: ArrayFactory[T] = new ArrayFactory[T]
 
-    val Array: ArrayFactory[T] = new ArrayFactory[T]
+    // Do not drop the following `toMap` (See https://github.com/scala/bug/issues/10496)
+    override def toStringMap[V](map: Map[T, V]): Map[String, V] = map.toMap
   }
 
 }
+
+/** ConcatenableMatchingString are non empty US-ASCII strings built with letters, digits,
+  * and some other (parameterizable) extra characters.
+  * We use them to represent identifiers. In this way, we avoid
+  * empty identifiers, escaping problems, and other similar pitfalls.
+  *
+  * ConcatenableMatchingString has the advantage over MatchingStringModule of being
+  * concatenable (and can be generate from number) without extra check.
+  * Those properties are heavily use to generate some ids by combining other existing
+  * ids.
+  */
+sealed abstract class ConcatenableMatchingStringModule extends StringModule {
+
+  def fromLong(i: Long): T
+
+  final def fromInt(i: Int): T = fromLong(i.toLong)
+
+  def concat(s: T, ss: T*): T
+}
+
+object ConcatenableMatchingStringModule {
+
+  def apply(
+      extraAllowedChars: Char => Boolean,
+      maxLength: Int = Int.MaxValue
+  ): ConcatenableMatchingStringModule =
+    new ConcatenableMatchingStringModule {
+      type T = String
+
+      override def fromString(s: String): Either[String, T] =
+        if (s.isEmpty)
+          Left(s"""empty string""")
+        else if (s.length > maxLength)
+          Left(s"""string too long""")
+        else
+          s.find(c => c > '\u007f' || !(c.isLetterOrDigit || extraAllowedChars(c)))
+            .fold[Either[String, T]](Right(s))(c =>
+              Left(s"""non expected character 0x${c.toInt.toHexString} in "$s""""))
+
+      override def fromLong(i: Long): T = i.toString
+
+      override def equalInstance: Equal[T] = scalaz.std.string.stringInstance
+
+      override val Array: ArrayFactory[T] = new ArrayFactory[T]
+
+      override def toStringMap[V](map: Map[T, V]): Map[String, V] = map
```

Comment:
Similarly to the other `toStringMap` here.

Reply:
done